File: src/algorithms/cpp/Coin_Change.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}


class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector <int> dp(amount + 1, amount + 1);
        dp[amount] = 0;
        for (int i = amount; i > 0; -- i) {
            for (int &coin: coins) {
                if (i - coin >= 0)
                    get_min(dp[i - coin], dp[i] + 1);
            }
        }
        return dp[0] == amount + 1 ? -1 : dp[0];
    }
};


int main() {
    Solution solution = Solution();
    vector <int> coins = {1, 2, 5};
    cout << solution.coinChange(coins, 11) << endl;
    return 0;
}

File: src/algorithms/python/Coin_Change.py
#!/bin/env python2
# -*- coding: utf-8 -*-


class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        dp = [0 if i == 0 else -1 for i in range(amount + 1)]
        for coin in coins:
            for i in range(coin, amount + 1):
                if (
                    dp[i - coin] != -1 and
                    (dp[i] == -1 or dp[i] > dp[i - coin] + 1)
                ):
                    dp[i] = dp[i - coin] + 1
        return dp[amount]
